libfwupdplugin/fu-cab-firmware.c

```diff
@@ -0,0 +1,792 @@
+/*
+ * Copyright (C) 2023 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN "FuCabFirmware"
+
+#include "config.h"
+
+#include <zlib.h>
+
+#include "fu-byte-array.h"
+#include "fu-bytes.h"
+#include "fu-cab-firmware.h"
+#include "fu-cab-image.h"
+#include "fu-cab-struct.h"
+#include "fu-chunk-array.h"
+#include "fu-mem-private.h"
+#include "fu-string.h"
+
+typedef struct {
+	gboolean compressed;
+	gboolean flattern;
+} FuCabFirmwarePrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE(FuCabFirmware, fu_cab_firmware, FU_TYPE_FIRMWARE)
+#define GET_PRIVATE(o) (fu_cab_firmware_get_instance_private(o))
+
+/**
+ * fu_cab_firmware_get_compressed:
+ * @self: a #FuCabFirmware
+ *
+ * Gets if the cabinet archive should be compressed.
+ *
+ * Returns: boolean
+ *
+ * Since: 1.9.7
+ **/
+gboolean
+fu_cab_firmware_get_compressed(FuCabFirmware *self)
+{
+	FuCabFirmwarePrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_IS_CAB_FIRMWARE(self), FALSE);
+	return priv->compressed;
+}
+
+/**
+ * fu_cab_firmware_set_compressed:
+ * @self: a #FuCabFirmware
+ * @compressed: boolean
+ *
+ * Sets if the cabinet archive should be compressed.
+ *
+ * Since: 1.9.7
+ **/
+void
+fu_cab_firmware_set_compressed(FuCabFirmware *self, gboolean compressed)
+{
+	FuCabFirmwarePrivate *priv = GET_PRIVATE(self);
+	g_return_if_fail(FU_IS_CAB_FIRMWARE(self));
+	priv->compressed = compressed;
+}
+
+/**
+ * fu_cab_firmware_get_flattern:
+ * @self: a #FuCabFirmware
+ *
+ * Gets if the cabinet archive filenames should be flatterned.
+ *
+ * Returns: boolean
+ *
+ * Since: 1.9.7
+ **/
+gboolean
+fu_cab_firmware_get_flattern(FuCabFirmware *self)
+{
+	FuCabFirmwarePrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_IS_CAB_FIRMWARE(self), FALSE);
+	return priv->flattern;
+}
+
+/**
+ * fu_cab_firmware_set_flattern:
```

Comment:
you mean flattened?  or is flattern something I don't understand?

Reply:
Hmm, i guess flattern is the verb -- what i'm trying to say is "do I ignore the path component and only take the basename?" -- gcab uses `nopath` for this -- other ideas welcome.

Reply:
how about `use_basename` and TRUE/FALSE

Reply:
counter-proposal `only_basename`?

Reply:
Great 😃

Reply:
Done.